src/Core/OrganizationFeatures/OrgUser/Invitation/Accept/OrganizationUserAcceptCommand.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Threading.Tasks;
+using Bit.Core.Models.Table;
+using Bit.Core.OrganizationFeatures.OrgUser.Mail;
+using Bit.Core.Repositories;
+using Bit.Core.Utilities;
+
+namespace Bit.Core.OrganizationFeatures.OrgUser.Invitation.Accept
+{
+    public class OrganizationUserAcceptCommand : IOrganizationUserAcceptCommand
+    {
+        private readonly IOrganizationUserAcceptAccessPolicies _organizationUserAcceptAccessPolicies;
+        private readonly IOrganizationRepository _organizationRepository;
+        private readonly IOrganizationUserRepository _organizationUserRepository;
+        private readonly IOrganizationUserInvitationService _organizationUserInvitationService;
+        private readonly IOrganizationUserMailer _organizationUserMailer;
+
+        public OrganizationUserAcceptCommand(
+            IOrganizationUserAcceptAccessPolicies organizationUserAcceptAccessPolicies,
+            IOrganizationRepository organizationRepository,
+            IOrganizationUserRepository organizationUserRepository,
+            IOrganizationUserInvitationService organizationUserInviteService,
+            IOrganizationUserMailer organizationUserMailer
+        )
+        {
+            _organizationUserAcceptAccessPolicies = organizationUserAcceptAccessPolicies;
+            _organizationRepository = organizationRepository;
+            _organizationUserRepository = organizationUserRepository;
+            _organizationUserInvitationService = organizationUserInviteService;
+            _organizationUserMailer = organizationUserMailer;
+        }
+
+        public async Task<OrganizationUser> AcceptUserAsync(Guid organizationUserId, User user, string token)
+        {
+            var orgUser = await _organizationUserRepository.GetByIdAsync(organizationUserId);
+            var org = await _organizationRepository.GetByIdAsync(orgUser.OrganizationId);
+
+            CoreHelpers.HandlePermissionResult(
+                await _organizationUserAcceptAccessPolicies.CanAcceptInviteAsync(org, user, orgUser, _organizationUserInvitationService.TokenIsValid(token, user, orgUser))
+            );
```

Comment:
It helps quite a bit for testing to simplify mocking. Separating the access concerns from execution allows us to chain them more easily and test them in a way we are actually sure they're working by just making sure a mock is called.

That being said, you're right, it makes circular dependencies easier to create. I never actually ran into it since they should only depend on the service and repository they represent.